### PR TITLE
[INLONG-7477][Sort] Fix the metadata of table write error for canal-json

### LINKED ISSUE #7477: [Bug][Sort] Fix the metadata of table write error for canal-json
### What happened

The metadata of table write error for canal-json.
image

### What you expected to happen

The metadata of table write right for canal-json.

### How to reproduce

It can reproduce using the follow flink sql 
```
CREATE TABLE `input`(
    PRIMARY KEY (`id`) NOT ENFORCED,
    `database_wedata_di` STRING METADATA FROM 'meta.database_name' VIRTUAL,
    `table_wedata_di` STRING METADATA FROM 'meta.table_name' VIRTUAL,
    `type_wedata_di` STRING METADATA FROM 'meta.op_type' VIRTUAL,
    `ts_wedata_di` TIMESTAMP_LTZ(3) METADATA FROM 'meta.op_ts' VIRTUAL,
    `processing_time_wedata_di` AS PROCTIME(),
    `id` INT,
    `name` STRING,
    `PROCESS_TIME` AS PROCTIME())
    WITH (
    'debezium.snapshot.locking.mode' = 'none',
    'scan.startup.mode' = 'initial',
    'debezium.binary.handling.mode' = 'base64',
    'inlong.metric.labels' = 'groupId=1&streamId=1&nodeId=1',
    'connector' = 'mysql-cdc-inlong',
    'hostname' = 'localhost',
    'database-name' = 'test',
    'port' = '3306',
    'scan.incremental.snapshot.enabled' = 'true',
    'server-time-zone' = 'Asia/Shanghai',
    'username' = 'root',
    'password' = 'password',
    'table-name' = 'test_db.test_table'
);
CREATE TABLE `output`(
    `id` INT,
    `name` STRING,
    `TABLE_NAME_wedata_di` STRING METADATA FROM 'value.table',
    `OP_TS_wedata_di` TIMESTAMP_LTZ(3) METADATA FROM 'value.event-timestamp',
    `DATABASE_NAME_wedata_di` STRING METADATA FROM 'value.database',
    `OP_TYPE_wedata_di` STRING METADATA FROM 'value.type')
    WITH (
    'dirty.side-output.field-delimiter' = ',',
    'dirty.side-output.format' = 'csv',
    'dirty.side-output.log-tag' = '脏数据][test_table',
    'dirty.identifier' = 'test_table',
    'dirty.side-output.enable' = 'true',
    'dirty.side-output.labels' = 'SYSTEM_TIME=${SYSTEM_TIME},test_table',
    'dirty.side-output.connector' = 'log',
    'dirty.ignore' = 'true',
    'inlong.metric.labels' = 'groupId=1&streamId=1&nodeId=2',
    'topic' = 'test_table',
    'properties.bootstrap.servers' = 'localhost:9092',
    'connector' = 'kafka-inlong',
    'canal-json-inlong.ignore-parse-errors' = 'true',
    'canal-json-inlong.map-null-key.mode' = 'DROP',
    'format' = 'canal-json-inlong',
    'canal-json-inlong.encode.decimal-as-plain-number' = 'true',
    'canal-json-inlong.timestamp-format.standard' = 'SQL',
    'canal-json-inlong.map-null-key.literal' = 'null'
);
INSERT INTO `output`
    SELECT 
    `id` AS `id`,
    `name` AS `name`,
    `table_wedata_di` AS `TABLE_NAME_wedata_di`,
    `ts_wedata_di` AS `OP_TS_wedata_di`,
    `database_wedata_di` AS `DATABASE_NAME_wedata_di`,
    `type_wedata_di` AS `OP_TYPE_wedata_di`
    FROM `input` ;
```

### Environment

_No response_

### InLong version

master

### InLong Component

InLong Sort

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by yunqingmoswu**
### Prepare a Pull Request
*(Change the title refer to the following example)*

- Title: [INLONG-7477][Sort] Fix the metadata of table write error for canal-json

*(The following *XYZ* should be replaced by the actual [GitHub Issue](https://github.com/apache/inlong/issues) number)*

- Fixes #7477

### Motivation

Fix the metadata of table write error for canal-json

### Modifications

Fix the metadata of table write error for canal-json in class 'CanalJsonEnhancedSerializationSchema.java'
image

### Verifying this change

*(Please pick either of the following options)*

- [ ] This change is a trivial rework/code cleanup without any test coverage.

- [x] This change is already covered by existing tests, such as:
  *(please describe tests)*

- [ ] This change added tests and can be verified as follows:

  *(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a follow-up issue for adding the documentation

### CODE REVIEW

**Review by healchow — APPROVED** [review]

**Review by dockerzhang — APPROVED** [review]

File: inlong-sort/sort-formats/format-json/src/main/java/org/apache/inlong/sort/formats/json/canal/CanalJsonEnhancedSerializationSchema.java [hunk 1]
```diff
@@ -122,9 +127,16 @@ private static RowType createJsonRowType(DataType physicalDataType, List<Writeab
 
     @Override
     public void open(InitializationContext context) {
-        reuse = new GenericRowData(2 + wirteableMetadataFieldGetter.length);
+        int size = 2 + wirteableMetadataFieldGetter.length;
+        if (typeIndex != -1) {
+            size--;
+        }
```
Comment (gong) [inline inlong-sort/sort-formats/format-json/src/main/java/org/apache/inlong/sort/formats/json/canal/CanalJsonEnhancedSerializationSchema.java:133]: I don't undestand this logic

Comment (yunqingmoswu) [reply inlong-sort/sort-formats/format-json/src/main/java/org/apache/inlong/sort/formats/json/canal/CanalJsonEnhancedSerializationSchema.java:133]: The metadata field  of 'type' has been written as the default value. At this time, the user displays the configured 'type', so the 'type' displayed by the user needs to be removed. At this time, the size of the final written data is reduced by 1.

**Review by gong — APPROVED** [review]

File: inlong-sort/sort-formats/format-json/src/main/java/org/apache/inlong/sort/formats/json/canal/CanalJsonEnhancedSerializationSchema.java (hunk 1 above)
Comment (EMsnap) [reply inlong-sort/sort-formats/format-json/src/main/java/org/apache/inlong/sort/formats/json/canal/CanalJsonEnhancedSerializationSchema.java:133]: please add a description for the method

File: inlong-sort/sort-formats/format-json/src/main/java/org/apache/inlong/sort/formats/json/canal/CanalJsonEnhancedSerializationSchema.java [hunk 2]
```diff
@@ -122,9 +127,16 @@ private static RowType createJsonRowType(DataType physicalDataType, List<Writeab
 
     @Override
     public void open(InitializationContext context) {
-        reuse = new GenericRowData(2 + wirteableMetadataFieldGetter.length);
+        int size = 2 + wirteableMetadataFieldGetter.length;
```
Comment (EMsnap) [inline inlong-sort/sort-formats/format-json/src/main/java/org/apache/inlong/sort/formats/json/canal/CanalJsonEnhancedSerializationSchema.java:130]: Please add a description for this for better understanding

**Review by EMsnap — APPROVED** [review]

**Pull request merged**